exchanges/deribit/deribit_websocket.go

```diff
@@ -642,140 +609,74 @@ func (d *Deribit) processCandleChart(respRaw []byte, channels []string) error {
-				return fmt.Errorf("%w, invalid amount", errMalformedData)
-			}
-			bids = append(bids, orderbook.Level{
-				Price:  price,
-				Amount: amount,
-			})
-		}
-		if len(asks) == 0 && len(bids) == 0 {
-			return nil
-		}
 
-		switch orderbookData.Type {
-		case "snapshot":
-			return d.Websocket.Orderbook.LoadSnapshot(&orderbook.Book{
-				Exchange:          d.Name,
-				ValidateOrderbook: d.ValidateOrderbook,
-				LastUpdated:       orderbookData.Timestamp.Time(),
-				Pair:              cp,
-				Asks:              asks,
-				Bids:              bids,
-				Asset:             a,
-				LastUpdateID:      orderbookData.ChangeID,
-			})
-		case "change":
-			return d.Websocket.Orderbook.Update(&orderbook.Update{
-				Asks:       asks,
-				Bids:       bids,
-				Pair:       cp,
-				Asset:      a,
-				UpdateID:   orderbookData.ChangeID,
-				UpdateTime: orderbookData.Timestamp.Time(),
-			})
-		}
-	} else if len(channels) == 5 {
-		cp, a, err := d.getAssetPairByInstrument(orderbookData.InstrumentName)
-		if err != nil {
-			return err
-		}
-		asks := make(orderbook.Levels, 0, len(orderbookData.Asks))
-		for x := range orderbookData.Asks {
-			if len(orderbookData.Asks[x]) != 2 {
-				return errMalformedData
-			}
-			price, okay := orderbookData.Asks[x][0].(float64)
-			if !okay {
-				return fmt.Errorf("%w, invalid orderbook price", errMalformedData)
-			} else if price == 0 {
-				continue
-			}
-			amount, okay := orderbookData.Asks[x][1].(float64)
-			if !okay {
-				return fmt.Errorf("%w, invalid amount", errMalformedData)
-			}
-			asks = append(asks, orderbook.Level{
-				Price:  price,
-				Amount: amount,
-			})
-		}
-		bids := make([]orderbook.Level, 0, len(orderbookData.Bids))
-		for x := range orderbookData.Bids {
-			if len(orderbookData.Bids[x]) != 2 {
-				return errMalformedData
-			}
-			price, okay := orderbookData.Bids[x][0].(float64)
-			if !okay {
-				return fmt.Errorf("%w, invalid orderbook price", errMalformedData)
-			} else if price == 0 {
+	buildLevels := func(items []orderbookItem) orderbook.Levels {
+		lvls := make(orderbook.Levels, 0, len(items))
+
+		for x := range items {
+			if items[x].Amount == 0 {
 				continue
 			}
-			amount, okay := orderbookData.Bids[x][1].(float64)
-			if !okay {
-				return fmt.Errorf("%w, invalid amount", errMalformedData)
-			}
-			bids = append(bids, orderbook.Level{
-				Price:  price,
-				Amount: amount,
+			lvls = append(lvls, orderbook.Level{
+				Price:  items[x].Price.Float64(),
+				Amount: items[x].Amount.Float64(),
 			})
 		}
-		if len(asks) == 0 && len(bids) == 0 {
-			return nil
-		}
-		return d.Websocket.Orderbook.LoadSnapshot(&orderbook.Book{
-			Asks:         asks,
-			Bids:         bids,
-			Pair:         cp,
-			Asset:        a,
-			Exchange:     d.Name,
-			LastUpdateID: orderbookData.ChangeID,
-			LastUpdated:  orderbookData.Timestamp.Time(),
+		return lvls
+	}
+	asks := buildLevels(ob.Asks)
+	bids := buildLevels(ob.Bids)
+	if len(asks) == 0 && len(bids) == 0 {
```

Comment:
This happening a lot? If you can check that the LastUpdateID is sequential even when no action has occurred on the books between the update frequency window we can actually load zero values in now.